### PR TITLE
[zero] optimize grad offload

### PR DESCRIPTION
**Pull request by ver217**
When `reuse_fp16_shard` is `True`, grad is stored in `p.col_attr.sharded_data_tensor.payload` and we move all payload to CPU. This reduces max GPU memory usage during backward.

### CODE REVIEW

File: colossalai/zero/sharded_optim/sharded_optim_v2.py [hunk 1]
```diff
@@ -161,7 +160,7 @@ def step(self, *args, **kwargs):
                 # Since p.data is fp32 and p.col_attr.sharded_data_tensor is fp16
 
                 # TODO() optimize this line CPU (fp32) -> GPU (fp16)
-                colo_model_data_tensor_move(p, p.col_attr.sharded_data_tensor)
+                p.col_attr.sharded_data_tensor.reset_payload(p.half().to(torch.cuda.current_device()))
```
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:163]: why not use colo_model_data_tensor_move_inline API.

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 2]
```diff
@@ -206,7 +207,7 @@ def _post_backward_operations(self) -> None:
             else:
                 grad_payload = cast_tensor_to_fp32(p.col_attr.fp16_grad)
             if p.col_attr.offload_grad:
-                grad_payload = colo_model_tensor_clone(grad_payload, torch.device('cpu'))
+                grad_payload.data = grad_payload.data.cpu()
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:210]: colo_model_data_tensor_move_inline?

File: colossalai/zero/sharded_optim/sharded_optim_v2.py (hunk 1 above)
Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py]: DONE

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 2 above)
Comment (ver217) [reply colossalai/zero/sharded_model/sharded_model_v2.py]: DONE

File: colossalai/zero/sharded_optim/sharded_optim_v2.py [hunk 3]
```diff
@@ -161,7 +161,8 @@ def step(self, *args, **kwargs):
                 # Since p.data is fp32 and p.col_attr.sharded_data_tensor is fp16
 
                 # TODO() optimize this line CPU (fp32) -> GPU (fp16)
-                colo_model_data_tensor_move(p, p.col_attr.sharded_data_tensor)
+                p.col_attr.sharded_data_tensor.reset_payload(
+                    colo_model_tensor_clone(p.half(), torch.cuda.current_device()))
```
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:165]: why not use colo_model_data_tensor_move?
```
colo_model_data_tensor_move(p.half().cuda(current_device(), p.col_attr.sharded_data_tensor)
```

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:165]: 1. When reuse fp16 shard and offload grad, `sharded_data_tensor` is on CPU. `colo_model_data_tensor_move` calls `tensor.copy_()` which won't change the device of `sharded_data_tensor` to CUDA.
2. According to the profile result of @1SAA , `tensor.copy_()` is slower than `tensor.half().to(torch.cuda.current_device())`. Besides, the latter one is easy to profile using our `PCIEProfiler`.

Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:208]: OK

**Review by feifeibear — APPROVED** [review]

**Pull request merged**